### ISSUE TITLE
$(DeviceSpecificOutputPath) doesn't take into account changes to $(OutputPath)

### ISSUE THREAD

**Issue by KirillOsenkov** [xamarin/xamarin-macios#12476]
`$(DeviceSpecificOutputPath)` is being set in Xamarin.Shared.props:
https://github.com/xamarin/xamarin-macios/blob/243cc4019f985ca4422cc038c6a86c23d6cc0e9f/msbuild/Xamarin.Shared/Xamarin.Shared.props#L254-L257

After the assignment reads the current value of `$(OutputPath)`, `$(OutputPath)` is usually mutated in `Microsoft.NET.TargetFrameworkInference.targets` to append the target framework:
https://github.com/dotnet/sdk/blob/7cade7d92a25695e24b4b5e6785415a2d372371e/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.TargetFrameworkInference.targets#L242

Since that is a .targets file, it's imported through Directory.Build.targets, which comes after Xamarin.Shared.props.

As a result, the `DeviceSpecificOutputPath` is set to the value without the target framework, resulting in a discrepancy where most files go under the target framework folder, but the raw and invalid .app is created one level above.

**Comment by KirillOsenkov** [comment]
I think moving the calculation of `DeviceSpecificOutputPath` to the Xamarin targets could fix this issue, but then the danger is that someone might attempt to read `DeviceSpecificOutputPath` before it's being set. We'd need to audit all the usages and ensure that they're all either in a target (i.e. not during the evaluation pass) or come after the `DeviceSpecificOutputPath` is set.

**Comment by KirillOsenkov** [comment]
Fortunately it doesn't look like there'll be many usages to audit:
![image](https://user-images.githubusercontent.com/679326/129963172-fdd8a5a3-d88e-497b-8be6-f388971b1210.png)

**Comment by KirillOsenkov** [comment]
Yes I think moving the usages from `Xamarin.Shared.props` into one of `Xamarin.Mac.Common.targets` or `Xamarin.Shared.targets` should work fine.

**Comment by rolfbjarne** [comment]
Is this is a .NET 6 project, or a legacy Xamarin.Mac project?

**Comment by KirillOsenkov** [comment]
Legacy Xamarin.Mac, SDK-style .csproj

**Comment by sandyarmstrong** [comment]
SDK-style comes from https://github.com/xamarin/mirepoix/tree/master/src/Xamarin.Mac.Sdk.

**Comment by rolfbjarne** [comment]
This isn't a problem in .NET.
